espflash/src/image_format/esp_idf.rs

```diff
@@ -817,9 +817,12 @@ where
 pub fn check_idf_bootloader(elf_data: &Vec<u8>) -> Result<()> {
     let object = File::parse(elf_data.as_slice()).into_diagnostic()?;
     let section = object.section_by_name(".rodata_desc").is_some();
+    let esp_hal = object.section_by_name(".espressif.metadata").is_some();
     let symbol = object.symbols().any(|sym| sym.name() == Ok("esp_app_desc"));
 
-    if !section || !symbol {
+    // esp-hal specific, because it searches for the hard-coded esp_app_desc symbol
+    // in an ELF section
+    if esp_hal && (!section || !symbol) {
         return Err(Error::AppDescriptorNotPresent).into_diagnostic();
```

Comment:
I think this bit still needs some more work, based on my comment here: https://github.com/esp-rs/espflash/issues/886#issuecomment-2975647493

We should still return an error if the section is missing from _either_ type of binary, but we should only tell users to use `esp-bootloader-idf` _if_ we detect that it's an esp-hal binary. I'm hoping we can customize the error context with `thiserror` or `miette` for this purpose :crossed_fingers: